pkg/prometheus/server/operator.go

```diff
@@ -972,6 +986,45 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 	return nil
 }
 
+func (c *Operator) shouldDelete(p *monitoringv1.Prometheus, sts *appsv1.StatefulSet) (bool, error) {
+	if !c.retentionPoliciesEnabled {
+		// Feature-gate is disabled, default behavior is always to delete.
+		return true, nil
+	}
+
+	srp := p.Spec.ShardRetentionPolicy
+	if srp == nil || srp.WhenScaledDown == monitoringv1.WhenScaledDownRetentionTypeDelete {
+		return true, nil
+	}
+
+	retentionAnnotation, ok := sts.Annotations[deletionTimestampAnnotation]
+	if !ok {
+		// This statefulset was not marked for deletion yet. Let's just add the annotation and return.
+		sts.Annotations[deletionTimestampAnnotation] = c.getRetentionDate(p)
+		return false, nil
+	}
+	retentionAsTime, err := time.Parse(time.RFC3339, retentionAnnotation)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse retention annotation %q: %w", deletionTimestampAnnotation, err)
+	}
+	if c.now().After(retentionAsTime) {
+		return true, nil
+	}
+
+	// We should never get here, but just in case we do, we return false for safety.
```

Comment:
you mean that because getRetentionDate() returns a very far away timestamp, we always return before? If yes, I'd remove the comment here to avoid a mismatch when we implement proper cleanup.